runtime/major_gc.c

```diff
@@ -343,30 +345,37 @@ void caml_shrink_mark_stack () {
    wasteful. Subsequent calls will continue progress.
  */
 static int redarken_chunk(char* heap_chunk, struct mark_stack* stk) {
-  value* p = Chunk_redarken_start(heap_chunk);
-  value* end = Chunk_redarken_end(heap_chunk);
-
-  while (p <= end) {
-    header_t hd = Hd_op(p);
-
-    if( Is_black_hd(hd) && Tag_hd(hd) < No_scan_tag ) {
-      if( stk->count < stk->size/4 ) {
-        mark_stack_push(stk, Val_op(p), 0, NULL);
-      } else {
-        /* Only fill up a quarter of the mark stack, we can resume later
-           for more if we need to */
-        Chunk_redarken_start(heap_chunk) = p;
-        return 0;
-      }
+  heap_chunk_head* chunk = Chunk_head(heap_chunk);
+  mark_entry me = chunk->redarken_first;
+  value* end = chunk->redarken_end;
+  while (me.end <= end) {
+    value v;
+    if (stk->count < stk->size/4) {
+      stk->stack[stk->count++] = me;
+    } else {
+      /* Only fill up a quarter of the mark stack, we can resume later
```

Comment:
Maybe comment about how you arrived at 1/4?

Reply:
I didn't! This logic is unchanged.